runtime/compiler/env/VMJ9.cpp

```diff
@@ -2347,6 +2350,113 @@ TR_J9VMBase::markHotField(TR::Compilation * comp, TR::SymbolReference * symRef,
+			if(NULL == newHotField) {
+				omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+				omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+				return; 
+			} 
+			newHotField->next = NULL;
+			newHotField->hotFieldOffset = fieldOffset;
+			newHotField->hotness = (reducedFrequency * reducedCpuUtil);
+			newHotField->cpuUtil = reducedCpuUtil;
+			omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+         
+         /* update hotFieldsInfo for the class now that a new hot field has been added */
+			clazz->hotFieldsInfo->hotFieldListHead = newHotField;
+			clazz->hotFieldsInfo->isClassHotFieldListDirty = true;
+			clazz->hotFieldsInfo->hotFieldListLength++;
+			clazz->hotFieldsInfo->classLoader = clazz->classLoader;
+			omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+			return;		
+		} else {
+			omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+		}
+	} 	
+
+	/* see if hot field exists already in the hot field list */
+	J9HotField* temp = clazz->hotFieldsInfo->hotFieldListHead;
+	while (NULL != temp) {
+      /* if the hot field exists, update the hot field with the newly reported hot field information */
+		if(temp->hotFieldOffset == fieldOffset) {
+			omrthread_monitor_enter(clazz->classLoader->hotFieldPoolMutex);
+			temp->hotness += (reducedFrequency * reducedCpuUtil);
+			temp->cpuUtil += reducedCpuUtil;
+			omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+			clazz->hotFieldsInfo->isClassHotFieldListDirty = true;
+			return; 
+		}
+      /* if the hot field does not exist and the hot field list is not currently at its max size, create and initialize the new hot field  */
+		else if(NULL == temp->next) {
+			if(clazz->hotFieldsInfo->hotFieldListLength < MAX_HOT_FIELD_LIST_LENGTH)
+			{
+				omrthread_monitor_enter(clazz->classLoader->hotFieldPoolMutex);
+				J9HotField* newHotField = (J9HotField *)pool_newElement(clazz->classLoader->hotFieldPool);
+				if(NULL == newHotField) {
+					omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+					return;
+				}
+				newHotField->next = NULL;
+				newHotField->hotFieldOffset = fieldOffset;
+				newHotField->hotness = (reducedFrequency * reducedCpuUtil);
+				newHotField->cpuUtil = reducedCpuUtil;
+				temp->next = newHotField;
+				omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+				/* update hotFieldsInfo for the class now that a new hot field has been added */
+            clazz->hotFieldsInfo->hotFieldListLength++;
+				clazz->hotFieldsInfo->isClassHotFieldListDirty = true;
+			} 				
+			return; 	
+		}
+		temp = temp->next;
+	}
+}
+
+/**
+ * Query if hot reference field is reqired for dynamicBreadthFirstScanOrdering
+ *  @return true if scavenger dynamicBreadthFirstScanOrdering is enabled, 0 otherwise 
+ */
+bool
+TR_J9VMBase::isHotReferenceFieldRequired()
+   {
+   return _jitConfig->javaVM->memoryManagerFunctions->j9gc_hot_reference_field_required(_jitConfig->javaVM);
```

Comment:
Perhaps this would be better done as a query on the objectmodel or something?

Reply:
Yes I believe a query on the object model would be more correct